extensions/lifecycle/tasks/LifecycleTask.js

```diff
@@ -71,67 +79,94 @@ class LifecycleTask extends BackbeatTask {
-                if (err) {
-                    log.error('error listing versioned bucket objects', {
-                        error: err,
-                        bucket: params.Bucket,
-                    });
-                    return next(err);
-                }
-                return next(null, data);
-            }),
-            (data, next) => {
-                if (data.IsTruncated) {
-                    // NOTE: this might only be needed if a LC rule defines
-                    //   a rule for processing NoncurrentVersion actions
-                    const entry = Object.assign({}, bucketData, {
-                        details: {
-                            keyMarker: data.NextKeyMarker,
-                            versionIdMarker: data.NextVersionIdMarker,
-                        },
-                    });
-                    return this.sendBucketEntry(entry, err => {
-                        if (!err) {
-                            log.debug(
-                                'sent kafka entry for bucket consumption', {
+    _getObjectVersions(bucketData, bucketLCRules, keys, prevDate, log, done) {
+        const params = {
+            Bucket: bucketData.target.bucket,
+            MaxKeys: MAX_KEYS,
+        };
+
+        async.eachLimit(keys, CONCURRENCY_DEFAULT, (key, cb) => {
+            if (bucketData.details.versionIdMarker &&
+            bucketData.details.keyMarker) {
+                params.KeyMarker = bucketData.details.keyMarker;
+                params.VersionIdMarker = bucketData.details.versionIdMarker;
+            }
+            params.Prefix = key;
+
+            const req = this.s3target.listObjectVersions(params);
+            attachReqUids(req, log);
+            async.waterfall([
+                next => req.send((err, data) => {
+                    if (err) {
+                        log.error('error listing versioned bucket objects', {
+                            error: err,
+                            bucket: params.Bucket,
+                        });
+                        return next(err);
+                    }
+                    return next(null, data);
+                }),
+                (data, next) => {
+                    // sorted newest to oldest
+                    // NOTE: staleDate can be undefined
+                    let staleDate = prevDate;
+
+                    const sortedVersions = data.Versions
+                        .reduce((store, version) => {
+                            // guarantee Key matches
+                            if (version.Key === key) {
+                                const copy = Object.assign({}, version,
+                                    { staleDate });
+                                staleDate = version.LastModified;
+                                store.push(copy);
+                            }
+                            return store;
+                        }, []);
+
+                    if (data.IsTruncated) {
```

Comment:
I think this applies to the current `listObjectVersions` of this particular object, I doubt you want to spawn a new bucket entry in case this particular listing is truncated (but you probably want to keep going and aggregate the rest of the listing as a single big list).

Reply:
I was worried this would potentially take a long time to complete as it would list versions for all keys of a bucket. Instead, I was thinking if `IsTruncated`, I get the `LastModified` date of the last version of the listing and pass this as `prevDate` to the bucket entry. 

On the next listing for that given key, even if the previous listings have all been expired already, I still have the `prevDate` to get the stale date for the next version in the listing.

Maybe I'm worrying about something that I don't need to worry about, but would love your input on this and can adjust and change as needed. 👍